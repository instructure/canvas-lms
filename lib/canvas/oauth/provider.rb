#
# Copyright (C) 2012 - present Instructure, Inc.
#
# This file is part of Canvas.
#
# Canvas is free software: you can redistribute it and/or modify it under
# the terms of the GNU Affero General Public License as published by the Free
# Software Foundation, version 3 of the License.
#
# Canvas is distributed in the hope that it will be useful, but WITHOUT ANY
# WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS FOR
# A PARTICULAR PURPOSE. See the GNU Affero General Public License for more
# details.
#
# You should have received a copy of the GNU Affero General Public License along
# with this program. If not, see <http://www.gnu.org/licenses/>.

module Canvas::Oauth
  class Provider
    OAUTH2_OOB_URI = 'urn:ietf:wg:oauth:2.0:oob'


    attr_reader :client_id, :redirect_uri, :scopes, :purpose

    def initialize(client_id, redirect_uri = "", scopes = [], purpose = nil)
      @client_id = client_id
      @redirect_uri = redirect_uri
      @scopes = scopes
      @purpose = purpose
    end

    def has_valid_key?
      key.present? && key.active?
    end

    def client_id_is_valid?
      return false unless @client_id.present?
      begin
        !!Integer(@client_id)
      rescue ArgumentError
        false
      end
    end

    def is_authorized_by?(secret)
      secret == key.api_key
    end

    def has_valid_redirect?
      self.class.is_oob?(redirect_uri) || key.redirect_domain_matches?(redirect_uri)
    end

    def icon_url
      key.icon_url
    end

    def key
      return nil unless client_id_is_valid?
      @key ||= DeveloperKey.where(id: @client_id).first
    end

    # Checks to see if a token has already been issued to this client and
    # if we can reissue the same token to that client without asking for
    # user permission again. If the developer key is trusted, access
    # tokens will be automatically authorized without prompting the end-
    # user
    def authorized_token?(user)
      if !self.class.is_oob?(redirect_uri)
        return true if Token.find_reusable_access_token(user, key, scopes, purpose)
        return true if key.trusted?
      end

      return false
    end

    def token_for(code)
      Token.new(key, code)
    end

    def token_for_refresh_token(refresh_token)
      access_token = AccessToken.authenticate_refresh_token(refresh_token)
      return nil unless access_token
      Token.new(key, nil, access_token)
    end

    def app_name
      key.name.presence || key.user_name.presence || key.email.presence || default_app_name
    end

    def redirect_uri
      @redirect_uri.presence || ""
    end

    def session_hash
      { :client_id => key.id, :redirect_uri => redirect_uri, :scopes => scopes, :purpose => purpose }
    end

    def self.is_oob?(uri)
      uri == OAUTH2_OOB_URI
    end

    def self.confirmation_redirect(controller, provider, logged_in_user)
      # skip the confirmation page if access is already (or automatically) granted
      if provider.authorized_token?(logged_in_user)
        final_redirect(controller, final_redirect_params(controller.session[:oauth2], logged_in_user))
      else
        controller.oauth2_auth_confirm_url
      end
    end

    def self.final_redirect_params(oauth_session, logged_in_user, options = {})
      options = {:scopes => oauth_session[:scopes], :remember_access => options[:remember_access], :purpose => oauth_session[:purpose]}
      code = Canvas::Oauth::Token.generate_code_for(logged_in_user.global_id, oauth_session[:client_id], options)
      redirect_params = { :code => code }
      redirect_params[:state] = oauth_session[:state] if oauth_session[:state]
      redirect_params
    end

    def self.final_redirect(controller, opts = {})
      session = controller.session
      redirect_uri = session[:oauth2][:redirect_uri]
      session.delete(:oauth2)

      if is_oob?(redirect_uri)
        controller.oauth2_auth_url(opts)
      else
        has_params = redirect_uri =~ %r{\?}
        redirect_uri + (has_params ? "&" : "?") + opts.to_query
      end
    end

    private
    def default_app_name
      I18n.translate('pseudonym_sessions.default_app_name', 'Third-Party Application')
    end


  end
end
