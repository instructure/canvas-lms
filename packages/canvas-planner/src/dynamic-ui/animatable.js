/*
 * Copyright (C) 2017 - present Instructure, Inc.
 *
 * This file is part of Canvas.
 *
 * Canvas is free software: you can redistribute it and/or modify it under
 * the terms of the GNU Affero General Public License as published by the Free
 * Software Foundation, version 3 of the License.
 *
 * Canvas is distributed in the hope that they will be useful, but WITHOUT ANY
 * WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS FOR
 * A PARTICULAR PURPOSE. See the GNU Affero General Public License for more
 * details.
 *
 * You should have received a copy of the GNU Affero General Public License along
 * with this program. If not, see <http://www.gnu.org/licenses/>.
 */

import React from 'react'
import {func, shape} from 'prop-types'

function getDisplayName(WrappedComponent) {
  return `Animatable(${WrappedComponent.displayName})`
}

// Components passed to registerAnimatable that the manager uses must provide this interface:
// - getFocusable() returns anything that has a `focus()` method, DOM element or otherwise
// - getScrollable() must return a DOM element
export function animatable(WrappedComponent) {
  return class Animatable extends React.Component {
    static displayName = getDisplayName(WrappedComponent)

    static contextTypes = {
      dynamicUiManager: shape({
        registerAnimatable: func,
        deregisterAnimatable: func
      })
    }

    registerAnimatable = (type, component, index, componentIds) => {
      // This should be required, but I don't want tests to have to muck with wrapping their stuff
      // in a DynamicUiProvider
      if (!this.context.dynamicUiManager) return
      this.context.dynamicUiManager.registerAnimatable(type, component, index, componentIds)
    }

    deregisterAnimatable = (type, component, componentIds) => {
      // This should be required, but I don't want tests to have to muck with wrapping their stuff
      // in a DynamicUiProvider
      if (!this.context.dynamicUiManager) return
      this.context.dynamicUiManager.deregisterAnimatable(type, component, componentIds)
    }

    render() {
      return (
        <WrappedComponent
          {...this.props}
          registerAnimatable={this.registerAnimatable}
          deregisterAnimatable={this.deregisterAnimatable}
        />
      )
    }
  }
}
