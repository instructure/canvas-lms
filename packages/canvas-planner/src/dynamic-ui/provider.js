/*
 * Copyright (C) 2017 - present Instructure, Inc.
 *
 * This file is part of Canvas.
 *
 * Canvas is free software: you can redistribute it and/or modify it under
 * the terms of the GNU Affero General Public License as published by the Free
 * Software Foundation, version 3 of the License.
 *
 * Canvas is distributed in the hope that they will be useful, but WITHOUT ANY
 * WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS FOR
 * A PARTICULAR PURPOSE. See the GNU Affero General Public License for more
 * details.
 *
 * You should have received a copy of the GNU Affero General Public License along
 * with this program. If not, see <http://www.gnu.org/licenses/>.
 */

import {Component, Children} from 'react'
import {shape, func, element} from 'prop-types'

const uiManagerShape = shape({
  handleAction: func.isRequired,
  registerAnimatable: func.isRequired,
  deregisterAnimatable: func.isRequired,
  preTriggerUpdates: func.isRequired,
  triggerUpdates: func.isRequired
})

export class DynamicUiProvider extends Component {
  static propTypes = {
    manager: uiManagerShape.isRequired,
    children: element.isRequired
  }

  static childContextTypes = {
    dynamicUiManager: uiManagerShape
  }

  constructor(props, context) {
    super(props, context)
    this.manager = props.manager
  }

  getChildContext() {
    return {
      dynamicUiManager: this.manager
    }
  }

  render() {
    return Children.only(this.props.children)
  }
}
