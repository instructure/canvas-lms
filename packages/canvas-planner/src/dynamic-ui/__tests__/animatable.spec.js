/*
 * Copyright (C) 2017 - present Instructure, Inc.
 *
 * This file is part of Canvas.
 *
 * Canvas is free software: you can redistribute it and/or modify it under
 * the terms of the GNU Affero General Public License as published by the Free
 * Software Foundation, version 3 of the License.
 *
 * Canvas is distributed in the hope that they will be useful, but WITHOUT ANY
 * WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS FOR
 * A PARTICULAR PURPOSE. See the GNU Affero General Public License for more
 * details.
 *
 * You should have received a copy of the GNU Affero General Public License along
 * with this program. If not, see <http://www.gnu.org/licenses/>.
 */

 import React from 'react';
 import {animatable} from '../animatable';
 import {DynamicUiProvider} from '../provider';
 import {mount} from 'enzyme';

class MockComponent extends React.Component {
  render () { return <div />; }
}

it('passes trigger property functions and forwards the calls to the dynamic ui manager', () => {
  const Wrapped = animatable(MockComponent);
  const mockManager = {
    handleAction: jest.fn(),
    registerAnimatable: jest.fn(),
    deregisterAnimatable: jest.fn(),
    preTriggerUpdates: jest.fn(),
    triggerUpdates: jest.fn(),
  };

  const wrapper = mount(
    <DynamicUiProvider manager={mockManager}>
      <Wrapped />
    </DynamicUiProvider>
  );
  expect(wrapper).toMatchSnapshot();

  const mockComponentProps = wrapper.find('MockComponent').props();
  mockComponentProps.registerAnimatable('type', 'component', 42, ['item']);
  expect(mockManager.registerAnimatable).toHaveBeenCalledWith('type', 'component', 42, ['item']);
  mockComponentProps.deregisterAnimatable('type', 'component', ['item']);
  expect(mockManager.deregisterAnimatable).toHaveBeenCalledWith('type', 'component', ['item']);
});
