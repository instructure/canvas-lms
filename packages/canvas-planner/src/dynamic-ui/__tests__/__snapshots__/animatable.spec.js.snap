// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`passes trigger property functions and forwards the calls to the dynamic ui manager 1`] = `
<DynamicUiProvider
  manager={
    Object {
      "deregisterAnimatable": [MockFunction],
      "handleAction": [MockFunction],
      "preTriggerUpdates": [MockFunction],
      "registerAnimatable": [MockFunction],
      "triggerUpdates": [MockFunction],
    }
  }
>
  <Animatable(MockComponent)>
    <MockComponent
      deregisterAnimatable={[Function]}
      registerAnimatable={[Function]}
    >
      <div />
    </MockComponent>
  </Animatable(MockComponent)>
</DynamicUiProvider>
`;
